Generate C++ demo and test files in build directory, rather than source directory

francesco-ballarin opened the pull request:
This PR changes CMakeLists.txt and form generation stages for C++ demos and tests so that the files are generated in the build directory and not in the source tree.

@RizzerOnGitHub I believe this will fix both #888 and #893. If possible start from a clean repo and a clean build dir.

The changes included here are not fully backward compatible. In particular, the demos and unittests target are not added anymore to the main makefile. Instead, demos and tests should be configured as standalone projects after the core library has been installed. CI has been updated to reflect this change. I believe that this separation is beneficial for CI testing, because these standalone projects (especially demos) were previously not covered, but could have been used by end users.

The only remaining build file to still be generated in the source dir is now dolfinx/common/version.h.

RizzerOnGitHub commented:
`cmake ..` with this branch is currently failing for my complex build:

    -- 
    -- Copying demo and test data to build directory.
    -- ----------------------------------------------
    -- 
    -- Generating form files in demo and test directories. May take some time...
    -- ----------------------------------------------------------------------------------------
    CMake Error at CMakeLists.txt:296 (message):
      Generation of form files failed:
    
      Compiling 1 forms in demo/poisson...
    
      usage: generate-form-files.py [-h] [--version] [-d] [-v] [-o
      OUTPUT_DIRECTORY]
    
                                    [--visualise] [-p] [-q QUADRATURE_RULE]
                                    [--quadrature-degree QUADRATURE_DEGREE]
                                    [-r {uflacs,tsfc}] [-f name value]
                                    [-u name value]
                                    ufl_file [ufl_file ...]
    
      generate-form-files.py: error: unrecognized arguments: --scalar_type
    
    
    
    -- Configuring incomplete, errors occurred!
    See also "/home/drew/projects/fenics/build/dolfinx/build-complex/obj-x86_64-linux-gnu-complex/CMakeFiles/CMakeOutput.log".


My real build was successful at the same point. My cmake configuration is the same for both builds (performed independently from fresh copies of the source) apart from setting PETSC_DIR and SLEPC_DIR.  Perhaps there's a problem in the configurations I've set up to get the complex build working (this configuration was successful against commit 797071f from 20200128).

My build configuration is at https://salsa.debian.org/science-team/fenics/dolfinx/-/tree/master/debian (in debian/rules, with a couple of patches in debian/patches)

RizzerOnGitHub commented:
The problem will be ll.45-46 in cpp/cmake/scripts/generate-form-files.py:

            if complex_mode:
                args += ["--scalar_type", "double complex"]

RizzerOnGitHub commented:
Yes, my build proceeds against current ffcx.

francesco-ballarin commented:
> Yes, my build proceeds against current ffcx.

Also tests and demos?

RizzerOnGitHub commented:
I don't have a complete build yet so haven't got the tests and demos confirmed yet.

The build is getting stuck building the python modules:

    [ 92%] Building CXX object CMakeFiles/cpp.dir/dolfinx/wrappers/refinement.cpp.o
    /usr/bin/c++  -DDOLFINX_VERSION=\"2019.2.9.99\" -DEIGEN_MAX_ALIGN_BYTES=32 -DHAS_PYBIND11_MPI4PY -DHAS_SLEPC -Dcpp_EXPORTS -I/usr/lib/petscdir/petsc3.12/x86_64-linux-gnu-real/lib/python3/dist-packages/petsc4py/include -I/usr/lib/python3/dist-packages/mpi4py/include -isystem /usr/include/python3.8 -isystem /usr/lib/python3/dist-packages/ffcx/codegeneration -isystem /usr/include/eigen3 -isystem /usr/include/hdf5/openmpi -isystem /usr/lib/x86_64-linux-gnu/openmpi/include/openmpi -isystem /usr/lib/x86_64-linux-gnu/openmpi/include -isystem /usr/lib/petscdir/petsc3.12/x86_64-linux-gnu-real/include -isystem /usr/lib/slepcdir/slepc3.12/x86_64-linux-gnu-real/include  -g -O2 -fdebug-prefix-map=/home/projects/fenics/build/dolfinx=. -fstack-protector-strong -Wformat -Werror=format-security -isystem /home/projects/fenics/build/dolfinx/debian/tmp-real/usr/include -DVERSION_INFO=\"2019.2.0.dev0\" -O3 -DNDEBUG -fPIC -fvisibility=hidden   -flto -fno-fat-lto-objects -Wall -Werror -pedantic -fvisibility=hidden -pthread -std=c++17 -o CMakeFiles/cpp.dir/dolfinx/wrappers/refinement.cpp.o -c /home/projects/fenics/build/dolfinx/build-real/python/dolfinx/wrappers/refinement.cpp
    In file included from /usr/include/c++/9/bits/move.h:55,
                     from /usr/include/c++/9/bits/stl_pair.h:59,
                     from /usr/include/c++/9/utility:70,
                     from /usr/include/c++/9/array:38,
                     from /home/projects/fenics/build/dolfinx/debian/tmp-real/usr/include/dolfinx/common/MPI.h:9,
                     from /home/projects/fenics/build/dolfinx/build-real/python/dolfinx/wrappers/MPICommWrapper.h:9,
                     from /home/projects/fenics/build/dolfinx/build-real/python/dolfinx/wrappers/caster_mpi.h:9,
                     from /home/projects/fenics/build/dolfinx/build-real/python/dolfinx/wrappers/mesh.cpp:7:
    /usr/include/c++/9/type_traits: In instantiation of ‘struct std::is_base_of<Eigen::MapBase<ompi_communicator_t, 0>, ompi_communicator_t>’:
    /usr/include/pybind11/detail/common.h:481:30:   required by substitution of ‘template<class Type> struct pybind11::detail::type_caster<Type, typename std::enable_if<std::integral_constant<bool, (decltype (pybind11::detail::is_template_base_of_impl<Eigen::EigenBase>::check((typename pybind11::detail::intrinsic_type<T>::type*)(nullptr)))::value && pybind11::detail::negation<std::integral_constant<bool, (std::integral_constant<bool, (decltype (pybind11::detail::is_template_base_of_impl<Eigen::DenseBase>::check((typename pybind11::detail::intrinsic_type<T>::type*)(nullptr)))::value && std::is_base_of<Eigen::MapBase<Derived, 0>, T>::value)>::value || (std::integral_constant<bool, (pybind11::detail::negation<std::integral_constant<bool, (decltype (pybind11::detail::is_template_base_of_impl<Eigen::DenseBase>::check((typename pybind11::detail::intrinsic_type<T>::type*)(nullptr)))::value && std::is_base_of<Eigen::MapBase<Derived, 0>, T>::value)> >::value && decltype (pybind11::detail::is_template_base_of_impl<Eigen::PlainObjectBase>::check((typename pybind11::detail::intrinsic_type<T>::type*)(nullptr)))::value)>::value || decltype (pybind11::detail::is_template_base_of_impl<Eigen::SparseMatrixBase>::check((typename pybind11::detail::intrinsic_type<T>::type*)(nullptr)))::value))> >::value)>::value, void>::type> [with Type = ompi_communicator_t]’
    /usr/include/c++/9/tuple:185:12:   required from ‘struct std::_Tuple_impl<0, pybind11::detail::type_caster<ompi_communicator_t, void>, pybind11::detail::type_caster<dolfinx::graph::AdjacencyList<long int>, void>, pybind11::detail::type_caster<dolfinx::fem::ElementDofLayout, void>, pybind11::detail::type_caster<Eigen::Ref<const Eigen::Array<double, -1, -1, 1, -1, -1>, 0, Eigen::OuterStride<-1> >, void> >’
    /usr/include/c++/9/tuple:523:11:   required from ‘class std::tuple<pybind11::detail::type_caster<ompi_communicator_t, void>, pybind11::detail::type_caster<dolfinx::graph::AdjacencyList<long int>, void>, pybind11::detail::type_caster<dolfinx::fem::ElementDofLayout, void>, pybind11::detail::type_caster<Eigen::Ref<const Eigen::Array<double, -1, -1, 1, -1, -1>, 0, Eigen::OuterStride<-1> >, void> >’
    /usr/include/pybind11/cast.h:1938:38:   required from ‘class pybind11::detail::argument_loader<ompi_communicator_t*, const dolfinx::graph::AdjacencyList<long int>&, const dolfinx::fem::ElementDofLayout&, const Eigen::Ref<const Eigen::Array<double, -1, -1, 1, -1, -1>, 0, Eigen::OuterStride<-1> >&>’
    /usr/include/pybind11/pybind11.h:133:50:   required from ‘void pybind11::cpp_function::initialize(Func&&, Return (*)(Args ...), const Extra& ...) [with Func = dolfinx::mesh::Mesh (*&)(ompi_communicator_t*, const dolfinx::graph::AdjacencyList<long int>&, const dolfinx::fem::ElementDofLayout&, const Eigen::Ref<const Eigen::Array<double, -1, -1, 1> >&); Return = dolfinx::mesh::Mesh; Args = {ompi_communicator_t*, const dolfinx::graph::AdjacencyList<long int>&, const dolfinx::fem::ElementDofLayout&, const Eigen::Ref<const Eigen::Array<double, -1, -1, 1, -1, -1>, 0, Eigen::OuterStride<-1> >&}; Extra = {pybind11::name, pybind11::scope, pybind11::sibling, char [37]}]’
    /usr/include/pybind11/pybind11.h:64:9:   required from ‘pybind11::cpp_function::cpp_function(Return (*)(Args ...), const Extra& ...) [with Return = dolfinx::mesh::Mesh; Args = {ompi_communicator_t*, const dolfinx::graph::AdjacencyList<long int>&, const dolfinx::fem::ElementDofLayout&, const Eigen::Ref<const Eigen::Array<double, -1, -1, 1, -1, -1>, 0, Eigen::OuterStride<-1> >&}; Extra = {pybind11::name, pybind11::scope, pybind11::sibling, char [37]}]’
    /usr/include/pybind11/pybind11.h:819:22:   required from ‘pybind11::module& pybind11::module::def(const char*, Func&&, const Extra& ...) [with Func = dolfinx::mesh::Mesh (*)(ompi_communicator_t*, const dolfinx::graph::AdjacencyList<long int>&, const dolfinx::fem::ElementDofLayout&, const Eigen::Ref<const Eigen::Array<double, -1, -1, 1> >&); Extra = {char [37]}]’
    /home/projects/fenics/build/dolfinx/build-real/python/dolfinx/wrappers/mesh.cpp:72:47:   required from here
    /usr/include/c++/9/type_traits:1301:12: error: invalid use of incomplete type ‘struct ompi_communicator_t’
     1301 |     struct is_base_of
          |            ^~~~~~~~~~
    In file included from /home/projects/fenics/build/dolfinx/debian/tmp-real/usr/include/dolfinx/common/MPI.h:20,
                     from /home/projects/fenics/build/dolfinx/build-real/python/dolfinx/wrappers/MPICommWrapper.h:9,
                     from /home/projects/fenics/build/dolfinx/build-real/python/dolfinx/wrappers/caster_mpi.h:9,
                     from /home/projects/fenics/build/dolfinx/build-real/python/dolfinx/wrappers/mesh.cpp:7:
    /usr/lib/x86_64-linux-gnu/openmpi/include/mpi.h:400:16: note: forward declaration of ‘struct ompi_communicator_t’
      400 | typedef struct ompi_communicator_t *MPI_Comm;
          |                ^~~~~~~~~~~~~~~~~~~
    In file included from /usr/include/c++/9/bits/move.h:55,
                     from /usr/include/c++/9/bits/stl_pair.h:59,
                     from /usr/include/c++/9/utility:70,
                     from /usr/include/c++/9/array:38,
                     from /home/projects/fenics/build/dolfinx/debian/tmp-real/usr/include/dolfinx/common/MPI.h:9,
                     from /home/projects/fenics/build/dolfinx/build-real/python/dolfinx/wrappers/MPICommWrapper.h:9,
                     from /home/projects/fenics/build/dolfinx/build-real/python/dolfinx/wrappers/caster_mpi.h:9,
                     from /home/projects/fenics/build/dolfinx/build-real/python/dolfinx/wrappers/mesh.cpp:7:

Have you seen this kind of problem with `is_base_of` or `ompi_communicator_t` before?

RizzerOnGitHub commented:
Disabling the python build, focusing just on the cpp test and demo, if I cmake the demo, the Makefile knows nothing of the actual demos, and does nothing.

        $ mkdir cpp/demo/rundemos/
    $ cd cpp/demo/rundemos/
    $ PETSC_DIR=/usr/lib/petsc SLEPC_DIR=/usr/lib/slepc cmake ..
    -- The C compiler identification is GNU 9.3.0
    -- The CXX compiler identification is GNU 9.3.0
    -- Check for working C compiler: /usr/bin/cc
    -- Check for working C compiler: /usr/bin/cc -- works
    -- Detecting C compiler ABI info
    -- Detecting C compiler ABI info - done
    -- Detecting C compile features
    -- Detecting C compile features - done
    -- Check for working CXX compiler: /usr/bin/c++
    -- Check for working CXX compiler: /usr/bin/c++ -- works
    -- Detecting CXX compiler ABI info
    -- Detecting CXX compiler ABI info - done
    -- Detecting CXX compile features
    -- Detecting CXX compile features - done
    -- Found MPI_C: /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so (found version "3.1") 
    -- Found MPI_CXX: /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi_cxx.so (found version "3.1") 
    -- Found MPI: TRUE (found version "3.1")  
    -- Found PkgConfig: /usr/bin/pkg-config (found version "0.29") 
    -- Checking for one of the modules 'craypetsc_real;PETSc'
    -- Looking for sys/types.h
    -- Looking for sys/types.h - found
    -- Looking for stdint.h
    -- Looking for stdint.h - found
    -- Looking for stddef.h
    -- Looking for stddef.h - found
    -- Check size of PetscInt
    -- Check size of PetscInt - failed
    -- Looking for PETSC_USE_COMPLEX
    -- Looking for PETSC_USE_COMPLEX - not found
    -- Checking for one of the modules 'crayslepc_real;SLEPc'
    -- Configuring done
    -- Generating done
    -- Build files have been written to: /home/drew/projects/fenics/build/dolfinx/build-real/cpp/demo/rundemos
    $ make all VERBOSE=1
    /usr/bin/cmake -S/home/drew/projects/fenics/build/dolfinx/build-real/cpp/demo -B/home/drew/projects/fenics/build/dolfinx/build-real/cpp/demo/rundemos --check-build-system CMakeFiles/Makefile.cmake 0
    /usr/bin/cmake -E cmake_progress_start /home/drew/projects/fenics/build/dolfinx/build-real/cpp/demo/rundemos/CMakeFiles /home/drew/projects/fenics/build/dolfinx/build-real/cpp/demo/rundemos/CMakeFiles/progress.marks
    make -f CMakeFiles/Makefile2 all
    make[1]: Entering directory '/home/drew/projects/fenics/build/dolfinx/build-real/cpp/demo/rundemos'
    make[1]: Nothing to be done for 'all'.
    make[1]: Leaving directory '/home/drew/projects/fenics/build/dolfinx/build-real/cpp/demo/rundemos'
    /usr/bin/cmake -E cmake_progress_start /home/drew/projects/fenics/build/dolfinx/build-real/cpp/demo/rundemos/CMakeFiles 0


If I try to build and run the unittests in cpp/test/unit, they do build against complex PETSc (set via PETSC_DIR), but they pick up the real build of dolfinx, libdolfinx_real.so.2019.2 instead of libdolfinx_complex.so.2019.2.  I think the challenge here is to manage cpp/test/unit/CMakeLists.txt with

    find_package(DOLFINX REQUIRED)

At the moment there's just one DOLFINXConfig.cmake installed (in /usr/share/dolfinx/cmake/).  I guess we need a mechanism to make that build-specific, i.e. allowing for real and complex builds installed simultaneously.  I've already patched other parts of the build to enable this (generating libdolfinx_real.so instead of libdolfinx.so), but the DOLFINX cmake file is where I've got stuck.  Simultaneous installation is a separate question from generated demo source files, so I'll raise a new Issue to discuss that.

francesco-ballarin commented:
> I don't have a complete build yet so haven't got the tests and demos confirmed yet.
> 
> The build is getting stuck building the python modules:
> 
> ```
> [ 92%] Building CXX object CMakeFiles/cpp.dir/dolfinx/wrappers/refinement.cpp.o
> /usr/bin/c++  -DDOLFINX_VERSION=\"2019.2.9.99\" -DEIGEN_MAX_ALIGN_BYTES=32 -DHAS_PYBIND11_MPI4PY -DHAS_SLEPC -Dcpp_EXPORTS -I/usr/lib/petscdir/petsc3.12/x86_64-linux-gnu-real/lib/python3/dist-packages/petsc4py/include -I/usr/lib/python3/dist-packages/mpi4py/include -isystem /usr/include/python3.8 -isystem /usr/lib/python3/dist-packages/ffcx/codegeneration -isystem /usr/include/eigen3 -isystem /usr/include/hdf5/openmpi -isystem /usr/lib/x86_64-linux-gnu/openmpi/include/openmpi -isystem /usr/lib/x86_64-linux-gnu/openmpi/include -isystem /usr/lib/petscdir/petsc3.12/x86_64-linux-gnu-real/include -isystem /usr/lib/slepcdir/slepc3.12/x86_64-linux-gnu-real/include  -g -O2 -fdebug-prefix-map=/home/projects/fenics/build/dolfinx=. -fstack-protector-strong -Wformat -Werror=format-security -isystem /home/projects/fenics/build/dolfinx/debian/tmp-real/usr/include -DVERSION_INFO=\"2019.2.0.dev0\" -O3 -DNDEBUG -fPIC -fvisibility=hidden   -flto -fno-fat-lto-objects -Wall -Werror -pedantic -fvisibility=hidden -pthread -std=c++17 -o CMakeFiles/cpp.dir/dolfinx/wrappers/refinement.cpp.o -c /home/projects/fenics/build/dolfinx/build-real/python/dolfinx/wrappers/refinement.cpp
> In file included from /usr/include/c++/9/bits/move.h:55,
>                  from /usr/include/c++/9/bits/stl_pair.h:59,
>                  from /usr/include/c++/9/utility:70,
>                  from /usr/include/c++/9/array:38,
>                  from /home/projects/fenics/build/dolfinx/debian/tmp-real/usr/include/dolfinx/common/MPI.h:9,
>                  from /home/projects/fenics/build/dolfinx/build-real/python/dolfinx/wrappers/MPICommWrapper.h:9,
>                  from /home/projects/fenics/build/dolfinx/build-real/python/dolfinx/wrappers/caster_mpi.h:9,
>                  from /home/projects/fenics/build/dolfinx/build-real/python/dolfinx/wrappers/mesh.cpp:7:
> /usr/include/c++/9/type_traits: In instantiation of ‘struct std::is_base_of<Eigen::MapBase<ompi_communicator_t, 0>, ompi_communicator_t>’:
> /usr/include/pybind11/detail/common.h:481:30:   required by substitution of ‘template<class Type> struct pybind11::detail::type_caster<Type, typename std::enable_if<std::integral_constant<bool, (decltype (pybind11::detail::is_template_base_of_impl<Eigen::EigenBase>::check((typename pybind11::detail::intrinsic_type<T>::type*)(nullptr)))::value && pybind11::detail::negation<std::integral_constant<bool, (std::integral_constant<bool, (decltype (pybind11::detail::is_template_base_of_impl<Eigen::DenseBase>::check((typename pybind11::detail::intrinsic_type<T>::type*)(nullptr)))::value && std::is_base_of<Eigen::MapBase<Derived, 0>, T>::value)>::value || (std::integral_constant<bool, (pybind11::detail::negation<std::integral_constant<bool, (decltype (pybind11::detail::is_template_base_of_impl<Eigen::DenseBase>::check((typename pybind11::detail::intrinsic_type<T>::type*)(nullptr)))::value && std::is_base_of<Eigen::MapBase<Derived, 0>, T>::value)> >::value && decltype (pybind11::detail::is_template_base_of_impl<Eigen::PlainObjectBase>::check((typename pybind11::detail::intrinsic_type<T>::type*)(nullptr)))::value)>::value || decltype (pybind11::detail::is_template_base_of_impl<Eigen::SparseMatrixBase>::check((typename pybind11::detail::intrinsic_type<T>::type*)(nullptr)))::value))> >::value)>::value, void>::type> [with Type = ompi_communicator_t]’
> /usr/include/c++/9/tuple:185:12:   required from ‘struct std::_Tuple_impl<0, pybind11::detail::type_caster<ompi_communicator_t, void>, pybind11::detail::type_caster<dolfinx::graph::AdjacencyList<long int>, void>, pybind11::detail::type_caster<dolfinx::fem::ElementDofLayout, void>, pybind11::detail::type_caster<Eigen::Ref<const Eigen::Array<double, -1, -1, 1, -1, -1>, 0, Eigen::OuterStride<-1> >, void> >’
> /usr/include/c++/9/tuple:523:11:   required from ‘class std::tuple<pybind11::detail::type_caster<ompi_communicator_t, void>, pybind11::detail::type_caster<dolfinx::graph::AdjacencyList<long int>, void>, pybind11::detail::type_caster<dolfinx::fem::ElementDofLayout, void>, pybind11::detail::type_caster<Eigen::Ref<const Eigen::Array<double, -1, -1, 1, -1, -1>, 0, Eigen::OuterStride<-1> >, void> >’
> /usr/include/pybind11/cast.h:1938:38:   required from ‘class pybind11::detail::argument_loader<ompi_communicator_t*, const dolfinx::graph::AdjacencyList<long int>&, const dolfinx::fem::ElementDofLayout&, const Eigen::Ref<const Eigen::Array<double, -1, -1, 1, -1, -1>, 0, Eigen::OuterStride<-1> >&>’
> /usr/include/pybind11/pybind11.h:133:50:   required from ‘void pybind11::cpp_function::initialize(Func&&, Return (*)(Args ...), const Extra& ...) [with Func = dolfinx::mesh::Mesh (*&)(ompi_communicator_t*, const dolfinx::graph::AdjacencyList<long int>&, const dolfinx::fem::ElementDofLayout&, const Eigen::Ref<const Eigen::Array<double, -1, -1, 1> >&); Return = dolfinx::mesh::Mesh; Args = {ompi_communicator_t*, const dolfinx::graph::AdjacencyList<long int>&, const dolfinx::fem::ElementDofLayout&, const Eigen::Ref<const Eigen::Array<double, -1, -1, 1, -1, -1>, 0, Eigen::OuterStride<-1> >&}; Extra = {pybind11::name, pybind11::scope, pybind11::sibling, char [37]}]’
> /usr/include/pybind11/pybind11.h:64:9:   required from ‘pybind11::cpp_function::cpp_function(Return (*)(Args ...), const Extra& ...) [with Return = dolfinx::mesh::Mesh; Args = {ompi_communicator_t*, const dolfinx::graph::AdjacencyList<long int>&, const dolfinx::fem::ElementDofLayout&, const Eigen::Ref<const Eigen::Array<double, -1, -1, 1, -1, -1>, 0, Eigen::OuterStride<-1> >&}; Extra = {pybind11::name, pybind11::scope, pybind11::sibling, char [37]}]’
> /usr/include/pybind11/pybind11.h:819:22:   required from ‘pybind11::module& pybind11::module::def(const char*, Func&&, const Extra& ...) [with Func = dolfinx::mesh::Mesh (*)(ompi_communicator_t*, const dolfinx::graph::AdjacencyList<long int>&, const dolfinx::fem::ElementDofLayout&, const Eigen::Ref<const Eigen::Array<double, -1, -1, 1> >&); Extra = {char [37]}]’
> /home/projects/fenics/build/dolfinx/build-real/python/dolfinx/wrappers/mesh.cpp:72:47:   required from here
> /usr/include/c++/9/type_traits:1301:12: error: invalid use of incomplete type ‘struct ompi_communicator_t’
>  1301 |     struct is_base_of
>       |            ^~~~~~~~~~
> In file included from /home/projects/fenics/build/dolfinx/debian/tmp-real/usr/include/dolfinx/common/MPI.h:20,
>                  from /home/projects/fenics/build/dolfinx/build-real/python/dolfinx/wrappers/MPICommWrapper.h:9,
>                  from /home/projects/fenics/build/dolfinx/build-real/python/dolfinx/wrappers/caster_mpi.h:9,
>                  from /home/projects/fenics/build/dolfinx/build-real/python/dolfinx/wrappers/mesh.cpp:7:
> /usr/lib/x86_64-linux-gnu/openmpi/include/mpi.h:400:16: note: forward declaration of ‘struct ompi_communicator_t’
>   400 | typedef struct ompi_communicator_t *MPI_Comm;
>       |                ^~~~~~~~~~~~~~~~~~~
> In file included from /usr/include/c++/9/bits/move.h:55,
>                  from /usr/include/c++/9/bits/stl_pair.h:59,
>                  from /usr/include/c++/9/utility:70,
>                  from /usr/include/c++/9/array:38,
>                  from /home/projects/fenics/build/dolfinx/debian/tmp-real/usr/include/dolfinx/common/MPI.h:9,
>                  from /home/projects/fenics/build/dolfinx/build-real/python/dolfinx/wrappers/MPICommWrapper.h:9,
>                  from /home/projects/fenics/build/dolfinx/build-real/python/dolfinx/wrappers/caster_mpi.h:9,
>                  from /home/projects/fenics/build/dolfinx/build-real/python/dolfinx/wrappers/mesh.cpp:7:
> ```
> 
> Have you seen this kind of problem with `is_base_of` or `ompi_communicator_t` before? Debian is in the middle of preparing the transition to gcc-10, I wonder if it's related to that.

I have never seen anything like this, but it is definitely unrelated (because this PR does not touch the python part). Is there a gcc-10 debian docker image I can try?

francesco-ballarin commented:
> if I cmake the demo, the Makefile knows nothing of the actual demos, and does nothing.

OK, I (and CI) do not test this, but is it really worth the effort? You are already in a build directory, what is the point of making a further build directory in a build directory?!

> Simultaneous installation is a separate question from generated demo source files, so I'll raise a new Issue to discuss that

OK!

RizzerOnGitHub commented:
> what is the point of making a further build directory in a build directory?

That would be an artifact of the build configuration I set up before you created this branch. I created the real and complex builds using separate copies of the toplevel source. I'll need to reorganise it to  test your branch fully (with two separate build dirs against the same source).

>this PR does not touch the python part). Is there a gcc-10 debian docker image I can try?

Understood. My python problem here must be something else. 

I haven't been using docker myself (the philosophy for Debian is to build the libraries and applications for the system as .deb packages).  But there seem to be Debian docker images at https://hub.docker.com/_/debian/   I'm using Debian unstable (sid), with fenicx from experimental, so you'd probably want the [experimental](https://github.com/debuerreotype/docker-debian-artifacts/blob/d6eeda93542f8e2a7d5f6e500b58fc4f12d055ce/experimental/Dockerfile) image (unless the required fenicx packages can be installed in an unstable/sid image after it's running).

francesco-ballarin commented:
> That would be an artifact of the build configuration I set up before you created this branch.

I have fixed this now (and rebased on master). The issue was using a relative path rather than absolute one.

francesco-ballarin commented:
@garth-wells I have requested a review from you because I think this PR still makes sense regardless of whether or not we are willing to fix in the short term each of #888, #893 and #901

RizzerOnGitHub commented:
Building and running demos for the real build passes well enough.

Building demos for the complex build comes into the problem that it links against the real build of libdolfinx.so (libdolfinx_real.so).  I think this the same issue that I mentioned about and raised in #901 about identifying the right DOLFINXConfig.cmake for the complex build to match the `find_package(DOLFINX REQUIRED)` in demo/CMakeLists.txt.

If you only have the one build then it works ok: the generated source files are in $BUILDDIR/demo/* and can be run with cmake after that build is done.

I gather your intention here is for the generated demo source files to be generated only when the library is being built, not to be generated independently later on (e.g. to test an existing installation).

RizzerOnGitHub commented:
I think I had the demo tests in the wrong order.  With your changes in this branch, you need the demos built *after* installation (previously I was running tests in the build dir before running the install step).

I can get the tests (both real and complex) to run after installation, though only if I specify the location of both  DOLFINXConfig.cmake and FindPETSc.cmake.  This works for me (for the complex test, launching from a sub dir of the build dir, as in "build-complex/demo/rundemo"):

    PETSC_DIR=/usr/lib/petscdir/petsc-complex SLEPC_DIR=/usr/lib/slepcdir/slepc-complex cmake  -DCMAKE_MODULE_PATH=`pwd`/../../../../debian/tmp-complex/usr/share/dolfinx/cmake -DDOLFINX_DIR=`pwd`/../../../../debian/tmp-complex/usr/share/dolfinx/cmake ..

where ../../../../debian/tmp-complex is the location is where I've "installed" the library (an intermediate destination ready for packaging), and needs to be specified by absolute path.  I had to add DOLFINX_DIR or else it couldn't find DOLFINXConfig.cmake (even though it's in the same CMAKE_MODULE_PATH).  I had to add CMAKE_MODULE_PATH or else it couldn't find FindPETSc.cmake, even though it's in DOLFINX_DIR. cmake is really annoying sometimes.

francesco-ballarin commented:
> I gather your intention here is for the generated demo source files to be generated only when the library is being built, not to be generated independently later on (e.g. to test an existing installation).

yes, I can confirm this. It seems to me that was the intention even before this PR, so I did not want to change that.

> I had to add DOLFINX_DIR
I think that could be expected, even though it never happened to me or CI (most probably because both of us have only one installation). For completeness, unlike CI, I never install in a system path in my configuration.

> I had to add CMAKE_MODULE_PATH or else it couldn't find FindPETSc.cmake
That I do not understand. 
That is supposed to be done already by
https://github.com/FEniCS/dolfinx/blob/master/cpp/cmake/templates/DOLFINXConfig.cmake.in#L30
Can you verify what path is hardcoded in your file at that line? Is it possible that somehow you move the installation away from that path (during the build script or because of a patch)?

RizzerOnGitHub commented:
My DOLFINXConfig.cmake.in in the toplevel source at cpp/cmake/templates/DOLFINXConfig.cmake.in has 

    list(APPEND CMAKE_MODULE_PATH "@CMAKE_INSTALL_PREFIX@/@CMAKE_INSTALL_DATAROOTDIR@/dolfinx/cmake")

same as the source on github.

In the installed version for the complex build, it renders as

    list(APPEND CMAKE_MODULE_PATH "/usr/share/dolfinx/cmake")

which is the final installation path, not the interim path used for packaging.  So this should be treated as a packaging issue, nothing to worry about further here.

francesco-ballarin commented:
ok, so to sum up this PR does indeed solve #888 and #893 for you, right? There are definitely other packaging issues that could possibly be discussed in future (starting from #901)

RizzerOnGitHub commented:
That's right, this PR places generated files under the build dir, so the real and complex demos can be built and run without crossing over on each other.

garth-wells commented:
@francesco-ballarin @RizzerOnGitHub is this PR ready to be merged?

francesco-ballarin commented:
Yes @garth-wells, if you agree on the changes.

RizzerOnGitHub commented:
@garth-wells It's passing my tests in the Debian build, I judge it ready.

francesco-ballarin commented:
@garth-wells this PR has been updated to also address the new pipeline introduced in #901. We believe it is ready for merge.

garth-wells commented on cpp/demo/CMakeLists.txt:
```diff
@@ -1,28 +1,20 @@
-  find_package(DOLFINX REQUIRED)
-endif()
+# Find DOLFINX config file
+find_package(DOLFINX REQUIRED)
```
Does this cause all the dolfinx CMake config to be performed even if it had already been computed by CMake?

francesco-ballarin replied:
It must be performed, `demo` folder is now configured as a standalone project.

garth-wells commented:
I didn't realise that this removes the `CMakeLists.txt` files from the demo sources directories.  This is a usability problem since users and developers can't navigate to the source, edit, build and run.

We should add back `CMakeLists.txt` files to the src.

RizzerOnGitHub commented:
The demo/CMakeLists.txt files are included in the installed files installed by 'make install'.  I'm seeing them intact in both the source and build dirs too after running build and install.
